cns/logger/cnslogger.go

```diff
@@ -0,0 +1,198 @@
+	c.logger.Close()
+	if c.th != nil {
+		c.th.Close(waitTimeInSecs)
+	}
+}
+
+func (c *CNSLogger) SetContextDetails(orchestrator, nodeID string) {
+	c.logger.Logf("SetContext details called with: %v orchestrator nodeID %v", orchestrator, nodeID)
+	c.m.Lock()
+	c.Orchestrator = orchestrator
+	c.NodeID = nodeID
+	c.m.Unlock()
+}
+
+func (c *CNSLogger) Printf(format string, args ...any) {
+	c.logger.Logf(format, args...)
+
+	if c.th == nil || c.DisableTraceLogging {
+		return
+	}
+
+	msg := fmt.Sprintf(format, args...)
+	c.sendTraceInternal(msg)
+}
+
+func (c *CNSLogger) Debugf(format string, args ...any) {
+	c.logger.Debugf(format, args...)
+
+	if c.th == nil || c.DisableTraceLogging {
+		return
+	}
+
+	msg := fmt.Sprintf(format, args...)
+	c.sendTraceInternal(msg)
+}
+
+func (c *CNSLogger) Errorf(format string, args ...any) {
+	c.logger.Errorf(format, args...)
+
+	if c.th == nil || c.DisableTraceLogging {
+		return
+	}
+
+	msg := fmt.Sprintf(format, args...)
+	c.sendTraceInternal(msg)
+}
+
+func (c *CNSLogger) Request(tag string, request any, err error) {
+	c.logger.Request(tag, request, err)
+
+	if c.th == nil || c.DisableTraceLogging {
+		return
+	}
+
+	var msg string
+	if err == nil {
+		msg = fmt.Sprintf("[%s] Received %T %+v.", tag, request, request)
+	} else {
+		msg = fmt.Sprintf("[%s] Failed to decode %T %+v %s.", tag, request, request, err.Error())
+	}
+
+	c.sendTraceInternal(msg)
+}
+
+func (c *CNSLogger) Response(tag string, response any, returnCode types.ResponseCode, err error) {
+	c.logger.Response(tag, response, int(returnCode), returnCode.String(), err)
+
+	if c.th == nil || c.DisableTraceLogging {
+		return
+	}
+
+	var msg string
+	switch {
+	case err == nil && returnCode == 0:
+		msg = fmt.Sprintf("[%s] Sent %T %+v.", tag, response, response)
+	case err != nil:
+		msg = fmt.Sprintf("[%s] Code:%s, %+v %s.", tag, returnCode.String(), response, err.Error())
+	default:
+		msg = fmt.Sprintf("[%s] Code:%s, %+v.", tag, returnCode.String(), response)
+	}
+
+	c.sendTraceInternal(msg)
+}
+
+func (c *CNSLogger) ResponseEx(tag string, request, response any, returnCode types.ResponseCode, err error) {
+	c.logger.ResponseEx(tag, request, response, int(returnCode), returnCode.String(), err)
+
+	if c.th == nil || c.DisableTraceLogging {
+		return
+	}
+
+	var msg string
+	switch {
+	case err == nil && returnCode == 0:
+		msg = fmt.Sprintf("[%s] Sent %T %+v %T %+v.", tag, request, request, response, response)
+	case err != nil:
+		msg = fmt.Sprintf("[%s] Code:%s, %+v, %+v, %s.", tag, returnCode.String(), request, response, err.Error())
+	default:
+		msg = fmt.Sprintf("[%s] Code:%s, %+v, %+v.", tag, returnCode.String(), request, response)
+	}
+
+	c.sendTraceInternal(msg)
+}
+
+func (c *CNSLogger) getOrchestratorAndNodeID() (orch, nodeID string) {
+	c.m.RLock()
+	orch, nodeID = c.Orchestrator, c.NodeID
+	c.m.RUnlock()
+	return
+}
+
+func (c *CNSLogger) sendTraceInternal(msg string) {
+	orch, nodeID := c.getOrchestratorAndNodeID()
```

Comment:
is this still racy, even with the mutex? I imagine we could lock, update, unlock, in two threads and when the first actually logs it will use the later value.
since that is current state, it might be out of scope to fix here...something that will be fixed as a side-effect of structured logging.

Reply:
yeah, I suppose this fixes the data race, not the possible race condition. I mentioned in a comment on the keyvault pr, I think the behavior really should be to lazily initialize these 2 variables (orchestrator and node id shouldn't become different for any CNS instance once set, is my understanding), but leaving that for a later PR.